### PR TITLE
Switch to using idna package

### PR DESCRIPTION
**Pull request by mahmoud**
As discussed in issue #19, Python's builtin idna support is outdated and broken. Thankfully the [idna](https://github.com/kjd/idna/) package exists. This PR switches to using that, instead of Python's builtin idna codec.

### CODE REVIEW

**Comment by glyph** [comment]
Thinking out loud my anxieties here, rather than a code review:
- I seem to remember that the built-in `idna` encoding manages to encode some things which are officially "invalid" yet recognized by browsers. Am I misremembering?
- Do we care about "zero dependencies"?  Years of being burned by packaging have left their mark; perhaps I shouldn't care, and if you can depend on hyperlink you can depend on its downstreams transitively.
- Why the ceiling pin in `install_requires`?  Is that just for semver compatibility?  Is there a process for noticing that the pin should float a version higher?  I'd be inclined to just leave it out, to avoid conflicting with other software that might need more recent versions.

**Comment by mahmoud** [comment]
@glyph, replies in order:

- I haven't reported/encountered anything like that. More worrisome is that idna encoding certain characters can result in totally-invalid URLs (slashes in the host, like #19). On the positive side, Cory Benfield says they use the idna package for everything. It could be there are old (pre-idna2k8) URLs lying around which might fail to decode, I can make it fallback. Note that I had to kill a doctest but it was clearly an example URL someone typed out, not linked to/used.
- I don't care all that much about no deps. Lots of hyper projects have dependencies. Pretty sure the Twisted `tls` extra pulls in idna anyways.
- The ceiling on install_requires was copied from requests and is basically semver at this point. I'm fine cutting that, too. The Twisted `tls` extra uses `idna!=2.3,>=0.6`.

File: hyperlink/_url.py [hunk 1]
```diff
@@ -1793,3 +1802,38 @@ def parse(url, decoded=True, lazy=False):
...
+* If a segment of a host (i.e., something in url.host.split('.')) is
+already ascii, idna doesn't perform its usual checks. For instance,
+capital letters are not valid idna2008. The package automatically lowercases.
+
+You'll get something like:
+
+> idna.core.InvalidCodepoint: Codepoint U+004B at position 1 ... not allowed
+
+This check and some other functionality can be bypassed by passing
+uts46=True to encode/decode. This allows a more permission and
```
Comment (wsanchez) [inline hyperlink/_url.py]: permission ➜ premissive

File: setup.py [hunk 2]
```diff
@@ -29,6 +29,7 @@
       zip_safe=False,
       license=__license__,
       platforms='any',
+      install_requires=['idna>=2.5,<2.7'],
```
Comment (wsanchez) [inline setup.py]: I would remove the `<2.7` ceiling

**Comment by mahmoud** [comment]
@wsanchez, all fixed up! The codecov integration on this PR is all froze up, but the report itself says the diff coverage is 100%, so I'm going to merge momentarily. Thanks!

**Comment by glyph** [comment]
Come on codecov do your thing.

**Comment by glyph** [comment]
(Please convince codecov that these lines are all covered somehow, then land.)

**Comment by mahmoud** [comment]
screenshot from 2018-01-07 16-59-46

@glyph yeah it's not gonna update, but the report on the site confirms 100% coverage, see screenshot above.

**Comment by wsanchez** [comment]
@glyph I think I'm not an approved approved because I'm not a committer.

**Pull request merged**